src/Diagnostics.DataProviders/KustoHeartBeatService/KustoHeartBeatService.cs

```diff
@@ -107,75 +113,97 @@ public KustoHeartBeat(string primaryCluster, string failoverCluster, KustoDataPr
+            {
+                UsePrimaryCluster = true;
+            }
 
-                }
-                catch (Exception ex)
+            LogHeartBeatInformation("Primary", primaryHeartBeatSuccess, PrimaryCluster, activityId, UsePrimaryCluster, exceptionForLog);
+        }
+
+        private async Task RunHeartBeatFailover(string activityId)
+        {
+            bool failoverHeartBeatSuccess = false;
+            Exception exceptionForLog = null;
+            try
+            {
+                var failoverHeartBeat = await _kustoDataProvider.ExecuteQueryForHeartbeat(_configuration.HeartBeatQuery, PrimaryCluster, _configuration.HeartBeatTimeOut, activityId);
+
+                if (failoverHeartBeat.Rows.Count >= 1)
                 {
-                    exceptionForLog = ex;
+                    failoverHeartBeatSuccess = true;
                 }
+            }
+            catch (Exception ex)
+            {
+                exceptionForLog = ex;
+            }
+
+            LogHeartBeatInformation("Failover", failoverHeartBeatSuccess, FailoverCluster, activityId, UsePrimaryCluster, exceptionForLog);
+        }
+
+        public async Task RunHeartBeatTask(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested) {
+                
+                string activityId = Guid.NewGuid().ToString();
 
-                if (primaryHeartBeatSuccess)
+                if (string.IsNullOrEmpty(FailoverCluster))
                 {
-                    _ConsecutiveFailureCount = 0;
-                    _ConsecutiveSuccessCount++;
+                    await RunHeartBeatPrimary(activityId);
                 }
                 else
                 {
-                    _ConsecutiveFailureCount++;
-                    _ConsecutiveSuccessCount = 0;
-                }
-
-                // if not in failover state
-                //  should failover?
-                if (UsePrimaryCluster && _ConsecutiveFailureCount >= _configuration.HeartBeatConsecutiveFailureLimit)
-                {
-                    UsePrimaryCluster = false;
-                } // else should stop failover
-                else if (!UsePrimaryCluster && _ConsecutiveSuccessCount >= _configuration.HeartBeatConsecutiveFailureLimit)
-                {
-                    UsePrimaryCluster = true;
+                    var primaryTask = RunHeartBeatPrimary(activityId);
+                    var failoverTask = RunHeartBeatFailover(activityId);
+                    await Task.WhenAll(new Task[] { primaryTask, failoverTask });
                 }
-
-                LogHeartBeatInformation(primaryHeartBeatSuccess, failoverHeartBeatSuccess, PrimaryCluster, FailoverCluster, activityId, _ConsecutiveSuccessCount, _ConsecutiveFailureCount, exceptionForLog);
-
                 await Task.Delay(TimeSpan.FromSeconds(_configuration.HeartBeatDelay), cancellationToken);
```

Comment:
Could we rename the configuration from `HeartBeatDelay` to `HeartBeatDelayInSeconds`. That way when we configure it in the future we don't have to look back to the code to check whether the delay is for seconds or milliseconds or minutes.